Improve specs and enforce consistency of container class docstring and default description

bendichter opened the pull request:
Depends on upcoming release of HDMF

bendichter committed:
Automatically generate all classes

Depends on upcoming release of HDMF

dsleiter committed:
Merge branch 'main' into auto_classes

dsleiter commented on src/spec/create_extension_spec.py:
```diff
@@ -58,7 +55,7 @@ def main():
         neurodata_type_def="NIRSSourcesTable",
         neurodata_type_inc="DynamicTable",
-        name="sources",
+        default_name="NIRSSourcesTable",
```
should this be considered a best practice for extensions - setting a `default_name` instead of a `name` and using the same value as `neurodata_type_inc`?

Looking through NWB schema docs, looks like this is most often but not always the case.

bendichter replied:
if you use "name" you can't have two of the objects in the same location. I prefer to use "default_name" so I have the flexibility to override if I need to.

dsleiter commented on src/spec/create_extension_spec.py:
```diff
@@ -91,12 +88,20 @@ def main():
             ),
         ],
+        attributes=[
+            NWBAttributeSpec(
```
`DynamicTable` already defines the `description` attribute - is the purpose of redefining here to set the `default_value`?

The `DynamicTable` class requires a `description` initialization parameter, so it seems that we always need to pass in a description value when instantiating a `NIRSSourcesTable` class built via `get_class`. Is there a way to use the `default_value` defined in the spec here? Or is the best way to define a custom `__init__` function and set the default `description` parameter to match the spec?

dsleiter commented on src/spec/create_extension_spec.py:
```diff
@@ -91,12 +88,20 @@ def main():
         neurodata_type_inc="DynamicTable",
-        name="detectors",
+        default_name="NIRSDetectorsTable",
         doc="A table describing optical detectors of a NIRS device",
```
Does the `doc` field for a spec have any impact on a class auto-built via `get_class`? Or is this just used when generating the docs for the schema (eg. https://nwb-schema.readthedocs.io/en/latest/format.html#sec-timeintervals)?

bendichter replied:
I know it is written to the schema and is a required field. It might be used as the docstring of the auto-generated class- I can't remember.

dsleiter committed:
Update tests to match class api of auto-generated classes

dsleiter reviewed (commented):
Thanks for the PR @bendichter! It was a great learning experience.

I went through and tested this. I had to make a few changes in order to get it running:
1. either define `__columns__` on auto-generated classes that inherit from `DynamicTable`, or change to reference columns using `table["col"]` rather than `table.col`
2. pass in a `description` parameter when instantiating table classes (because it is a required parameter of `DynamicTable`)
3. use column names in the singular when using `target_tables` in `NIRSChannelsTable` (as defined in the spec)

Are there benefits from using auto-generation besides not having to create the class manually?

I want to keep some of the benefits from original custom classes like custom `__init__`, setting `__columns__`. If auto-generation has additional benefits, then I might do that and then attach the extras afterward. Otherwise, it might be more explicit just to fully define the custom class.

I also asked a few other questions on the PR to better understand decisions and best practices.

dsleiter commented:
@bendichter I tried to push the fixes I mentioned, but it was denied. Could you please select "Allow edits from maintainers"?

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Let me know if you don't have the checkbox available as shown in that link, you might be running into the same issue I have when contributing to pynwb from a fork owned by an organization. I have in my notes somewhere how to use the github webapi  as a workaround to set that field.

bendichter commented:
@dsleiter those adjustments sound right.

For the matter of if there are advantages to doing it this way, the only real advantage is having less code and therefor less room for error. I generally encourage new users to do it automatically, because the manual way can require a lot of background knowledge about docval and how NWB works internally. However, specifying it manually has the advantage that it allows you to make adjustments in the mapping between the API and the neurodata type. I like having this relationship be deterministic and predictable, but that may become a problem, particularly if you want to change the schema while keeping the API the same. The automatic way does have the ability to do some custom input validation, as demonstrated [here](https://pynwb.readthedocs.io/en/stable/extensions_tutorial/4_auto_api.html#customizing-automatically-generated-apis).

bendichter commented:
I have added you as a contributor to my branch of the extension. If that doesn't work, another approach might be to pull from my branch and edit it yourself in your own fork

dsleiter committed:
Backout auto-generation of classes derived from DynamicTable.

* the auto-generated classes required either changing the user API of
  these classes or significant monkey-patching
* we decided to keep the full class definitions as they maintain the
  simplified user API and eliminates the need for monkey-patching

dsleiter committed:
Update neurodata type spec and verify consistency with container classes

* update neurodatatype spec doc fields to match the default value for the
  description attribute
* add tests to verify that the default description in initializers match
  the default value listed in the spec
* add tests to verify that container docstrings are consistent with the
  neurodatatype doc field

dsleiter committed:
Remove redundant tests

dsleiter committed:
Update release notes and credits

dsleiter commented:
@bendichter thanks again for the PR. We decided not to use auto-class generation for the dynamic tables in order to continue using the customized class behavior without needing a bunch of monkey patching. We do continue using the auto-generation for other types though.

You also made some improvements to the specs, so we want to keep those. Based on our discussion above, I also added some tests to verify consistency between the spec and the container class in the following ways:
* the class docstring should match or at least start with the value of the 'doc' field in the neurodatatype spec.
* the default value for the description in the class initializer should match the 'default_value' field of the description attribute in the spec

If you have the time and desire, could you please take a look at the updates to the PR?

dsleiter committed:
Merge branch 'main' into auto_classes

bendichter reviewed (commented):
looks good to me!
